File: maths/greatest_common_divisor.js
/**
 * Euclidean GCD
 */

function greatestCommonDivisorRecursive(a, b) {
    if (a === 0) return Math.abs(b);
    return greatestCommonDivisorRecursive(b % a, a);
}

function greatestCommonDivisorIterative(a, b) {
    while (b !== 0) {
        let temp = b;
        b = a % b;
        a = temp;
    }
    return Math.abs(a);
}

const a = 24;
const b = 40;
console.log(`GCD Recursive (${a}, ${b}) = ${greatestCommonDivisorRecursive(a, b)}`);
console.log(`GCD Iterative (${a}, ${b}) = ${greatestCommonDivisorIterative(a, b)}`);

File: maths/greatest_common_divisor.rs
pub fn greatest_common_divisor_recursive(a: i32, b: i32) -> i32 {
    if a == 0 {
        return b.abs();
    }
    greatest_common_divisor_recursive(b % a, a)
}

pub fn greatest_common_divisor_iterative(mut a: i32, mut b: i32) -> i32 {
    while b != 0 {
        let temp = b;
        b = a % b;
        a = temp;
    }
    a.abs()
}

fn main() {
    let a = 24;
    let b = 40;
    println!("GCD Recursive ({}, {}) = {}", a, b, greatest_common_divisor_recursive(a, b));
    println!("GCD Iterative ({}, {}) = {}", a, b, greatest_common_divisor_iterative(a, b));
}
